### ISSUE TITLE
[Help Wanted] [PLT-200] Allow hosting of MM at any route like http://www.example.com/chat

### ISSUE THREAD

**Issue by esethna** [mattermost/mattermost-server#7490]
If you're interested please comment here and come [join our "Contributors" community channel](https://pre-release.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://pre-release.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://docs.mattermost.com/guides/developer.html), specifically for [machine setup](https://docs.mattermost.com/developer/developer-setup.html) and for [developer workflow](https://docs.mattermost.com/developer/developer-flow.html).

----

**Notes**: [Jira ticket](https://mattermost.atlassian.net/browse/PLT-200)

Related to https://github.com/mattermost/platform/issues/643
Also request from [@botherder](https://twitter.com/botherder): https://twitter.com/botherder/status/655394442313138176
Voting: https://mattermost.uservoice.com/forums/306457-general/suggestions/12468372-install-mattermost-in-a-subdirectory
Forum issue: https://forum.mattermost.org/t/how-do-i-set-the-base-url-for-apache-ssl-forwarding/573

To achieve this change, the client should reference the [Site URL](|https://docs.mattermost.com/administration/config-settings.html#site-url) config.json setting instead of a relative link, when one is available.  If no site URL is set, then return a slash or a blank value.

Notes:
1) For someone to host Mattermost in a subdirectory, it is assumed that Site URL is required to be set.
2) This change should not break existing deployments, including those who don't have Site URL set.
3) Preliminary work from [~akihikodaki] in 2016 can be found here: https://github.com/mattermost/mattermost-server/compare/master...akihikodaki:PLT-200

**Comment by moonmeister** [comment]
I can probably take this. I'll take a look later see if it looks like something I can tackle.

**Comment by jasonblais** [comment]
Awesome!! 👍

**Comment by jasonblais** [comment]
A discussion related to the ticket spurred up in the last 30 minutes, so it might be worth reviewing that before starting on this: https://pre-release.mattermost.com/core/pl/ry91wiu7r7rs8qwdkh68bstqty

**Comment by moonmeister** [comment]
@jasonblais I'll definitely take this.

**Comment by moonmeister** [comment]
This is a massive change that will touch a lot of things. If anyone wants to see the progress: https://github.com/moonmeister/platform/tree/PLT-200

**Comment by moonmeister** [comment]
I want to take a second to confirm the scope of this ticket and what I'm targeting given the discussion @jasonblais mentioned above. 

1. The go `platform` executable will continue bind to the port and address configured in `ServiceSettings.ListenAddress`. 
1. We want to support reverse proxies that host MM in a subdirectory.

1. For things that require a static url reference (see below), `siteURL` must be set (for now, this could be fixed in another PR as it's out of scope here).
    * OAuth
    * Email Links
    * ???
1. The `siteURL` being set should not keep a user from accessing it from a different domain.
1.  The `siteURL` will effectively always contains a path, whether that path is `/` or `/chat`. Whatever the path is set to must be consistent across all domains.

Thoughts? Anything I missed @hmhealey or @jwilander ?

**Comment by jwilander** [comment]
Sounds about right to me

**Comment by moonmeister** [comment]
As an aside this is theoretically already possible using a finely tuned Reverse Proxy for any one who doesn't want to wait. See [here](http://www.apachetutor.org/admin/reverseproxies)

**Comment by moonmeister** [comment]
I'm struggling on this one in a addition to not having a ton of time. My understanding of JS and React is limiting. I think I have most of the Go/Server side done and some of the webapp side done. If anyone wants to/can help let me know.

**Comment by jasonblais** [comment]
@moonmeister Are there particular areas you're looking for help?

**Comment by moonmeister** [comment]
I have added some data parsing for the url and binding port/address on the go side but I need this data available in react. @hmhealey recommended using a redux store but I have 0 understanding of redux and limited React experience. So, I probably most need help in making that data available so that when building urls in JS/React I have the path data available. Thanks.

**Comment by jasonblais** [comment]
@moonmeister Would you be interested working on the Redux store? If so, I can ping a developer to help give you some direction.

If you'd prefer someone else to continue, an option would be to submit PRs with `WIP` in the title so others know it is a work-in-progress, and I can take a look if someone in the community was interested continuing from them.

**Comment by moonmeister** [comment]
Yeah, I'm happy to try.

**Comment by jasonblais** [comment]
That's great! @moonmeister posted a question here: https://pre-release.mattermost.com/core/pl/yjd78b6jibbdipf3ciy6fubtcr

**Comment by robydev** [comment]
Hello all. is this still under development? I really need this feature. Let me know if I can help.

**Comment by jasonblais** [comment]
@robydev We'd love your help. If you're interested, let me know and I can prepare a summary of past discussion here to help you get started.

**Comment by robydev** [comment]
ok. how can I start looking at the code? fork, pull request or branch?

**Comment by jasonblais** [comment]
Great! There was some progress made by @moonmeister at https://github.com/moonmeister/platform/tree/PLT-200

There is also some discussion here that might be relevant: https://pre-release.mattermost.com/core/pl/74s6ns9reibozbnrngasxqeyry

**Comment by robydev** [comment]
Hi Jason,

I'm struggling with development. I can't get a maintainable development environment.

What IDE are you using? How are you debbuging the service? Any unittests?

**Comment by jasonblais** [comment]
Hey @rbdevr, have you taken a look at the developer machine setup guide yet? https://docs.mattermost.com/developer/dev-setup.html

**Comment by jasonblais** [comment]
This is now completed and scheduled for Mattermost v5.1

**Comment by rbdevr** [comment]
Awesome job Jason. Please create a document explaining how to setup MM server to work eith subdirectories. 
Great job!
